File: src/go/problems/num_decodings_test.go
// https://leetcode.com/problems/decode-ways

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type NumDecodings struct{}

func TestNumDecodings(t *testing.T) {
	gen := core.TestSuite[NumDecodings]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("12").Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param("226").Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param("06").Result(0)
	}).Run(t)
}

func (NumDecodings) Solution(s string) int {
	temp1 := 1
	temp2 := 1
	for i := len(s) - 1; i >= 0; i-- {
		temp := temp1

		if s[i] == '0' {
			temp1 = 0
		} else if i+1 < len(s) && (s[i] == '1' || s[i] == '2' && s[i+1] <= '6') {
			temp1 += temp2
		}

		temp2 = temp
	}

	return temp1
}

File: src/python/challenges/problems/num_decodings_test.py
#https://leetcode.com/problems/decode-ways/
from core.problem_base import *

class NumDecodings(ProblemBase):
    def Solution(self, s: str) -> int:
        n = len(s)
        temp1 = 1
        temp2 = 1
        for i in range(n - 1, -1, -1):
            temp = temp1
            if s[i] == '0':
                temp1 = 0
            elif i + 1 < n and (s[i] == '1' or s[i] == '2' and s[i + 1] <= '6'):
                temp1 += temp2

            temp2 = temp

        return temp1


if __name__ == '__main__':
    TestGen(NumDecodings) \
        .Add(lambda tc: tc.Param("12").Result(2)) \
        .Add(lambda tc: tc.Param("226").Result(3)) \
        .Add(lambda tc: tc.Param("06").Result(0)) \
        .Run()
